feature: Add support for Testable Examples

### Did you check docs and existing issues?

- [x] I have read the [documentation](https://github.com/fredrikaverpil/neotest-golang/blob/main/README.md).
- [x] I have searched the existing [issues](https://github.com/fredrikaverpil/neotest-golang/issues).

### Is your feature request related to a problem? Please describe.

When you define [example functions](https://go.dev/blog/examples) (-> see 'Additional context' below if unfarmiliar with the feature) in your _test.go files, neotest-golang indicates that it runs them together with the normal test functions and indicates their status afterwards.
When you execute the whole file or start the watcher, neotest-golang indicates that a test example function passes, even though a wrong `// Output:` comment is provided.
Only when you run the example test function using the 'run nearest' or 'run all tests' functionality, does the function fail, as it is supposed to.

Example:

I implement a simple Add() function:

```
// Package add implements the Add function
package add

// Add takes two integers and returns their sum.
func Add(a, b int) int {
	return a + b
}
```

I create a add_test.go file that implements a normal test function and a testable example for Add() that declares a wrong `Output:` (0 instead of 8):

```
package add

import (
	"fmt"
	"testing"
)

func TestAdd(t *testing.T) {
	got := Add(2, 2)
	want := 4

	if got != want {
		t.Errorf("got %d, want %d", got, want)
	}
}

func ExampleAdd() {
	sum := Add(4, 4)
	fmt.Println(sum)
	// Output:
	// 0
}
```

I run `go test` on the add_test.go file which indicates a failure of the example:

```
❯ go test -json ./add/...
{"Time":"2025-11-20T17:02:23.279039581+01:00","Action":"start","Package":"github.com/malte-brunst/learn_go_with_tests/add"}
{"Time":"2025-11-20T17:02:23.284032736+01:00","Action":"run","Package":"github.com/malte-brunst/learn_go_with_tests/add","Test":"TestAdd"}
{"Time":"2025-11-20T17:02:23.284187964+01:00","Action":"output","Package":"github.com/malte-brunst/learn_go_with_tests/add","Test":"TestAdd","Output":"=== RUN   TestAdd\n"}
{"Time":"2025-11-20T17:02:23.284220462+01:00","Action":"output","Package":"github.com/malte-brunst/learn_go_with_tests/add","Test":"TestAdd","Output":"--- PASS: TestAdd (0.00s)\n"}
{"Time":"2025-11-20T17:02:23.284238372+01:00","Action":"pass","Package":"github.com/malte-brunst/learn_go_with_tests/add","Test":"TestAdd","Elapsed":0}
{"Time":"2025-11-20T17:02:23.284251224+01:00","Action":"run","Package":"github.com/malte-brunst/learn_go_with_tests/add","Test":"ExampleAdd"}
{"Time":"2025-11-20T17:02:23.284256282+01:00","Action":"output","Package":"github.com/malte-brunst/learn_go_with_tests/add","Test":"ExampleAdd","Output":"=== RUN   ExampleAdd\n"}
{"Time":"2025-11-20T17:02:23.284263695+01:00","Action":"output","Package":"github.com/malte-brunst/learn_go_with_tests/add","Test":"ExampleAdd","Output":"--- FAIL: ExampleAdd (0.00s)\n"}
{"Time":"2025-11-20T17:02:23.284269253+01:00","Action":"output","Package":"github.com/malte-brunst/learn_go_with_tests/add","Test":"ExampleAdd","Output":"got:\n"}
{"Time":"2025-11-20T17:02:23.284275244+01:00","Action":"output","Package":"github.com/malte-brunst/learn_go_with_tests/add","Test":"ExampleAdd","Output":"8\n"}
{"Time":"2025-11-20T17:02:23.284280281+01:00","Action":"output","Package":"github.com/malte-brunst/learn_go_with_tests/add","Test":"ExampleAdd","Output":"want:\n"}
{"Time":"2025-11-20T17:02:23.284288791+01:00","Action":"output","Package":"github.com/malte-brunst/learn_go_with_tests/add","Test":"ExampleAdd","Output":"0\n"}
{"Time":"2025-11-20T17:02:23.284293794+01:00","Action":"fail","Package":"github.com/malte-brunst/learn_go_with_tests/add","Test":"ExampleAdd","Elapsed":0}
{"Time":"2025-11-20T17:02:23.284298364+01:00","Action":"output","Package":"github.com/malte-brunst/learn_go_with_tests/add","Output":"FAIL\n"}
{"Time":"2025-11-20T17:02:23.285317156+01:00","Action":"output","Package":"github.com/malte-brunst/learn_go_with_tests/add","Output":"FAIL\tgithub.com/malte-brunst/learn_go_with_tests/add\t0.005s\n"}
{"Time":"2025-11-20T17:02:23.285356221+01:00","Action":"fail","Package":"github.com/malte-brunst/learn_go_with_tests/add","Elapsed":0.006}
```

I use neotest-golang instead and use the `run test (leader + t + t in LazyVim)` command.
A green tik indicates the functions succeeded.
The same wrong status is reported when you use the `watcher mode (leader + t + w in LazyVim)`.
However, neotest-golang reports the correct status (FAIL), when I use the `run all tests (leader + t + T in LazyVim)` or the `run nearest test command (leader + t + r)`.

### Describe the solution you'd like to see.

I would like neotest-golang to report the correct test status (pass/fail) for example functions in the _test.go files.
The correct status should be reported no matter which execution mode (watcher, run whole file, run all tests, run nearest test) is chosen.
The correct status is the one provided when using the standard `go test` command.

### Describe alternatives you've considered.

There are no alternatives.

### Additional context

As with typical tests, examples are functions that reside in a package’s _test.go files. Unlike normal test functions, though, example functions take no arguments and begin with the word Example instead of Test.
Example from the [Go example repository](https://cs.opensource.google/go/x/example):

```
package reverse_test

import (
    "fmt"

    "golang.org/x/example/hello/reverse"
)

func ExampleString() {
    fmt.Println(reverse.String("hello"))
    // Output: olleh
}
```

Running the package’s test suite, we can see the example function is executed with no further arrangement from us:

```
$ go test -v
=== RUN   TestString
--- PASS: TestString (0.00s)
=== RUN   ExampleString
--- PASS: ExampleString (0.00s)
PASS
ok      golang.org/x/example/hello/reverse  0.209s
```

As it executes the example, the testing framework captures data written to standard output and then compares the output against the example’s “Output:” comment. The test passes if the test’s output matches its output comment or failes otherwise.

Godoc uses a naming convention to associate an example function with a package-level identifier.

```
func ExampleFoo()     // documents the Foo function or type
func ExampleBar_Qux() // documents the Qux method of type Bar
func Example()        // documents the package as a whole
```

Following this convention, godoc displays the ExampleString example alongside the documentation for the String function.

Multiple examples can be provided for a given identifier by using a suffix beginning with an underscore followed by a lowercase letter. Each of these examples documents the String function:

```
func ExampleString()
func ExampleString_second()
func ExampleString_third()
```

--- a/lua/neotest-golang/runspec/file.lua
+++ b/lua/neotest-golang/runspec/file.lua
@@ -148,7 +148,7 @@ function M.get_regexp(filepath)
   local regexp = nil
   local lines = {}
   for line in io.lines(filepath) do
-    if line:match("func Test") then
+    if line:match("func Test") or line:match("func Example") then
       line = line:gsub("func ", "")
       line = line:gsub("%(.*", "")
       table.insert(lines, lib.convert.to_gotest_regex_pattern(line))

--- a/spec/integration/examples_spec.lua
+++ b/spec/integration/examples_spec.lua
@@ -1,0 +1,342 @@
+local _ = require("plenary")
+local options = require("neotest-golang.options")
+local path = require("neotest-golang.lib.path")
+
+-- Load integration helpers
+local integration_path = vim.uv.cwd() .. "/spec/helpers/integration.lua"
+local integration = dofile(integration_path)
+
+describe("Integration: Example functions", function()
+  it(
+    "file reports test discovery and execution for Example functions",
+    function()
+      -- ===== ARRANGE =====
+      local test_options = options.get()
+      test_options.runner = "gotestsum"
+      options.set(test_options)
+
+      local position_id = vim.uv.cwd()
+        .. "/tests/go/internal/examples/examples_test.go"
+      position_id = path.normalize_path(position_id)
+
+      -- Expected complete adapter execution result
+      ---@type AdapterExecutionResult
+      local want = {
+        results = {
+          -- Package-level result (from streaming) - should fail due to failing examples
+          [path.get_directory(position_id)] = {
+            status = "failed",
+            errors = {},
+          },
+          -- File-level result - should fail due to failing examples
+          [position_id] = {
+            status = "failed",
+            errors = {},
+          },
+          -- Regular test result
+          [position_id .. "::TestAdd"] = {
+            status = "passed",
+            errors = {},
+          },
+          -- Package-level Example (should pass)
+          [position_id .. "::Example"] = {
+            status = "passed",
+            errors = {},
+          },
+          -- Example function that passes
+          [position_id .. "::ExampleAdd"] = {
+            status = "passed",
+            errors = {},
+          },
+          -- Example function that fails (incorrect output comment)
+          [position_id .. "::ExampleAdd_failing"] = {
+            status = "failed",
+            errors = {},
+          },
+          -- Example for Multiply (should pass)
+          [position_id .. "::ExampleMultiply"] = {
+            status = "passed",
+            errors = {},
+          },
+          -- Second example for Multiply (should pass)
+          [position_id .. "::ExampleMultiply_second"] = {
+            status = "passed",
+            errors = {},
+          },
+          -- Example with method-style naming (should fail)
+          [position_id .. "::ExampleCalculator_Add"] = {
+            status = "failed",
+            errors = {},
+          },
+        },
+        run_spec = {
+          command = {}, -- this will be replaced in the assertion
+          context = {
+            pos_id = position_id,
+          },
+        },
+        strategy_result = {
+          code = 1, -- Exit code 1 because some examples fail
+        },
+        tree = {
+          -- this will be replaced in the assertion
+          _children = {},
+          _nodes = {},
+          _key = function()
+            return ""
+          end,
+        },
+      }
+
+      -- ===== ACT =====
+      ---@type AdapterExecutionResult
+      local got = integration.execute_adapter_direct(position_id)
+
+      -- ===== ASSERT =====
+      -- Copy dynamic run_spec fields
+      want.run_spec.command = got.run_spec.command
+      want.run_spec.cwd = got.run_spec.cwd
+      want.run_spec.env = got.run_spec.env
+      want.run_spec.stream = got.run_spec.stream
+      want.run_spec.strategy = got.run_spec.strategy
+      want.run_spec.context.golist_data = got.run_spec.context.golist_data
+      want.run_spec.context.stop_filestream =
+        got.run_spec.context.stop_filestream
+      want.run_spec.context.test_output_json_filepath =
+        got.run_spec.context.test_output_json_filepath
+
+      -- Copy dynamic strategy_result fields
+      want.strategy_result.output = got.strategy_result.output
+
+      -- Copy tree field if present
+      want.tree = got.tree
+
+      -- Copy dynamic output paths for all results
+      for pos_id, result in pairs(got.results) do
+        if want.results[pos_id] then
+          -- Copy output path if it exists
+          if result.output then
+            want.results[pos_id].output = result.output
+          end
+          -- Copy short field if it exists
+          if result.short then
+            want.results[pos_id].short = result.short
+          end
+        end
+      end
+
+      -- Helper function to sort errors for order-agnostic comparison
+      local function sort_errors(errors)
+        if not errors or #errors == 0 then
+          return errors or {}
+        end
+        local sorted = vim.deepcopy(errors)
+        table.sort(sorted, function(a, b)
+          if a.line ~= b.line then
+            return a.line < b.line
+          end
+          return a.message < b.message
+        end)
+        return sorted
+      end
+
+      -- Sort errors in both expected and actual results for order-agnostic comparison
+      for pos_id, result in pairs(want.results) do
+        if result.errors then
+          result.errors = sort_errors(result.errors)
+        end
+      end
+      for pos_id, result in pairs(got.results) do
+        if result.errors then
+          result.errors = sort_errors(result.errors)
+        end
+      end
+
+      assert.are.same(
+        vim.inspect(want),
+        vim.inspect(got),
+        "Complete adapter execution result should match"
+      )
+    end
+  )
+
+  it("failing Example function executes correctly", function()
+    -- ===== ARRANGE =====
+    local test_options = options.get()
+    test_options.runner = "gotestsum"
+    options.set(test_options)
+
+    local position_id = vim.uv.cwd()
+      .. "/tests/go/internal/examples/examples_test.go::ExampleAdd_failing"
+    position_id = path.normalize_path(position_id)
+    local file_id = vim.uv.cwd()
+      .. "/tests/go/internal/examples/examples_test.go"
+    file_id = path.normalize_path(file_id)
+
+    -- Expected complete adapter execution result
+    ---@type AdapterExecutionResult
+    local want = {
+      results = {
+        -- Package-level result (from streaming)
+        [path.get_directory(position_id)] = {
+          status = "failed",
+          errors = {},
+        },
+        -- File-level result
+        [file_id] = {
+          status = "failed",
+          errors = {},
+        },
+        -- Individual failing example
+        [position_id] = {
+          status = "failed",
+          errors = {},
+        },
+      },
+      run_spec = {
+        command = {},
+        context = {
+          pos_id = position_id,
+        },
+      },
+      strategy_result = {
+        code = 1, -- Exit code 1 for failing test
+      },
+      tree = {
+        _children = {},
+        _nodes = {},
+        _key = function()
+          return ""
+        end,
+      },
+    }
+
+    -- ===== ACT =====
+    ---@type AdapterExecutionResult
+    local got = integration.execute_adapter_direct(position_id)
+
+    -- ===== ASSERT =====
+    -- Copy dynamic fields
+    want.run_spec.command = got.run_spec.command
+    want.run_spec.cwd = got.run_spec.cwd
+    want.run_spec.env = got.run_spec.env
+    want.run_spec.stream = got.run_spec.stream
+    want.run_spec.strategy = got.run_spec.strategy
+    want.run_spec.context.golist_data = got.run_spec.context.golist_data
+    want.run_spec.context.stop_filestream = got.run_spec.context.stop_filestream
+    want.run_spec.context.test_output_json_filepath =
+      got.run_spec.context.test_output_json_filepath
+    want.run_spec.context.process_test_results =
+      got.run_spec.context.process_test_results
+    want.strategy_result.output = got.strategy_result.output
+    want.tree = got.tree
+
+    -- Copy dynamic output paths
+    for pos_id, result in pairs(got.results) do
+      if want.results[pos_id] then
+        if result.output then
+          want.results[pos_id].output = result.output
+        end
+        if result.short then
+          want.results[pos_id].short = result.short
+        end
+      end
+    end
+
+    assert.are.same(
+      vim.inspect(want),
+      vim.inspect(got),
+      "Failing Example function should be detected correctly"
+    )
+  end)
+
+  it("passing Example function executes correctly", function()
+    -- ===== ARRANGE =====
+    local test_options = options.get()
+    test_options.runner = "gotestsum"
+    options.set(test_options)
+
+    local position_id = vim.uv.cwd()
+      .. "/tests/go/internal/examples/examples_test.go::ExampleAdd"
+    position_id = path.normalize_path(position_id)
+    local file_id = vim.uv.cwd()
+      .. "/tests/go/internal/examples/examples_test.go"
+    file_id = path.normalize_path(file_id)
+
+    -- Expected complete adapter execution result
+    ---@type AdapterExecutionResult
+    local want = {
+      results = {
+        -- Package-level result (from streaming)
+        [path.get_directory(position_id)] = {
+          status = "passed",
+          errors = {},
+        },
+        -- File-level result
+        [file_id] = {
+          status = "passed",
+          errors = {},
+        },
+        -- Individual passing example
+        [position_id] = {
+          status = "passed",
+          errors = {},
+        },
+      },
+      run_spec = {
+        command = {},
+        context = {
+          pos_id = position_id,
+        },
+      },
+      strategy_result = {
+        code = 0, -- Exit code 0 for passing test
+      },
+      tree = {
+        _children = {},
+        _nodes = {},
+        _key = function()
+          return ""
+        end,
+      },
+    }
+
+    -- ===== ACT =====
+    ---@type AdapterExecutionResult
+    local got = integration.execute_adapter_direct(position_id)
+
+    -- ===== ASSERT =====
+    -- Copy dynamic fields
+    want.run_spec.command = got.run_spec.command
+    want.run_spec.cwd = got.run_spec.cwd
+    want.run_spec.env = got.run_spec.env
+    want.run_spec.stream = got.run_spec.stream
+    want.run_spec.strategy = got.run_spec.strategy
+    want.run_spec.context.golist_data = got.run_spec.context.golist_data
+    want.run_spec.context.stop_filestream = got.run_spec.context.stop_filestream
+    want.run_spec.context.test_output_json_filepath =
+      got.run_spec.context.test_output_json_filepath
+    want.run_spec.context.process_test_results =
+      got.run_spec.context.process_test_results
+    want.strategy_result.output = got.strategy_result.output
+    want.tree = got.tree
+
+    -- Copy dynamic output paths
+    for pos_id, result in pairs(got.results) do
+      if want.results[pos_id] then
+        if result.output then
+          want.results[pos_id].output = result.output
+        end
+        if result.short then
+          want.results[pos_id].short = result.short
+        end
+      end
+    end
+
+    assert.are.same(
+      vim.inspect(want),
+      vim.inspect(got),
+      "Passing Example function should be detected correctly"
+    )
+  end)
+end)

--- a/tests/go/internal/examples/examples.go
+++ b/tests/go/internal/examples/examples.go
@@ -1,0 +1,20 @@
+// Package examples provides simple math operations for testing Example functions
+package examples
+
+// Add returns the sum of two integers
+func Add(a, b int) int {
+	return a + b
+}
+
+// Multiply returns the product of two integers
+func Multiply(a, b int) int {
+	return a * b
+}
+
+// Calculator is a simple calculator type
+type Calculator struct{}
+
+// Add method on Calculator
+func (c Calculator) Add(a, b int) int {
+	return Add(a, b)
+}

--- a/tests/go/internal/examples/examples_test.go
+++ b/tests/go/internal/examples/examples_test.go
@@ -1,0 +1,57 @@
+package examples
+
+import (
+	"fmt"
+	"testing"
+)
+
+// Regular test to ensure Tests and Examples can coexist
+func TestAdd(t *testing.T) {
+	result := Add(2, 2)
+	if result != 4 {
+		t.Errorf("Add(2, 2) = %d; want 4", result)
+	}
+}
+
+// Package-level example (should pass)
+func Example() {
+	fmt.Println("examples package")
+	// Output: examples package
+}
+
+// Example for Add function (should pass)
+func ExampleAdd() {
+	sum := Add(2, 2)
+	fmt.Println(sum)
+	// Output: 4
+}
+
+// Example that should fail (incorrect output comment)
+func ExampleAdd_failing() {
+	sum := Add(3, 3)
+	fmt.Println(sum)
+	// Output: 5
+}
+
+// Example for Multiply function (should pass)
+func ExampleMultiply() {
+	product := Multiply(3, 4)
+	fmt.Println(product)
+	// Output: 12
+}
+
+// Example with multiple outputs (should pass)
+func ExampleMultiply_second() {
+	fmt.Println(Multiply(2, 3))
+	fmt.Println(Multiply(4, 5))
+	// Output:
+	// 6
+	// 20
+}
+
+// Example that demonstrates method-style naming (should fail)
+func ExampleCalculator_Add() {
+	result := Add(10, 5)
+	fmt.Println(result)
+	// Output: 100
+}

